### PR TITLE
Fix types for _error page

### LINKED ISSUE #7927: TS ErrorComponent => getInitialProps not async
# Bug report

## Describe the bug

Trying to port my TypeScript Next 8 to Next 9 and get typecheck inside the error component 

I am extending the generic error component with my own and I need to do some async operations inside `getInitialProps`. `@types/next` has `getInitialProps` can return a promise not not the new built in types. Is there a reason?

Error:
```
Class static side 'typeof CustomError' incorrectly extends base class static side 'typeof Error'.
  Types of property 'getInitialProps' are incompatible.
    Type '({ res, err }: NextPageContext) => Promise<{ statusCode: number; }>' is not assignable to type '({ res, err }: NextPageContext) => { statusCode: number | undefined; }'.
      Property 'statusCode' is missing in type 'Promise<{ statusCode: number; }>' but required in type '{ statusCode: number | undefined; }'.ts(2417)
```

## To Reproduce

```typescript
import ErrorComponent from 'next/error';

class CustomError extends ErrorComponent {
  	static async getInitialProps({ res }: NextPageContext) {
      	if (res.statusCode !== 500) {
			await doSomething();
     	}
		return { statusCode }
  	}
}

```

## Expected behavior

I assume that `getInitialProps` is async like any other page.

## Screenshots

N/A

## System information

- OS: macOS
- Browser (if applies): chrome
- Version of Next.js: 9.0.1

### PR DESCRIPTION
**Pull request by huv1k**
Fixes #7927

### CODE REVIEW

**Review by Timer — CHANGES_REQUESTED** [review]
This will break the sync variant, no? We need to represent both options.

**Review by timneutkens — DISMISSED** [review]
This should also be compatible with sync usage

File: packages/next/pages/_error.tsx [hunk 1]
```diff
@@ -20,7 +20,7 @@ export type ErrorProps = {
 export default class Error<P = {}> extends React.Component<P & ErrorProps> {
   static displayName = 'ErrorPage'
 
-  static getInitialProps({ res, err }: NextPageContext) {
+  static async getInitialProps({ res, err }: NextPageContext) {
```
Comment (kachkaev) [inline packages/next/pages/_error.tsx]: Not sure why `async` is required here given https://github.com/zeit/next.js/pull/7763

https://github.com/zeit/next.js/blob/be3eab2ad7892ca9a5be38f133494e700875804a/packages/next-server/lib/utils.ts#L16

**Review by Timer — APPROVED** [review]
LGTM

File: test/integration/typescript/pages/_error.tsx [hunk 2]
```diff
@@ -0,0 +1,11 @@
+import { NextPageContext } from 'next'
+import ErrorComponent from 'next/error'
+
+class CustomError extends ErrorComponent {
+  static async getInitialProps({ res }: NextPageContext) {
```
Comment (kachkaev) [inline test/integration/typescript/pages/_error.tsx]: This function is also sync by nature. Might be worth reflecting this in some future PR.

Comment (Timer) [reply test/integration/typescript/pages/_error.tsx]: It's probably best to make it `async` so we have that flexibility in the future IMO.

Comment (huv1k) [reply test/integration/typescript/pages/_error.tsx]: Making new PR for ti 😇

**Pull request merged**